falcon/middleware.py

```diff
@@ -42,38 +45,53 @@ class CORSMiddleware(object):
     ):
-        if allow_origins == '*':
-            self.allow_origins = allow_origins
+        self.allow_origins_wildcard = False
+        if isinstance(allow_origins, Callable):
+            self.allow_origins_filter = allow_origins
         else:
-            if isinstance(allow_origins, str):
-                allow_origins = [allow_origins]
-            self.allow_origins = frozenset(allow_origins)
-            if '*' in self.allow_origins:
-                raise ValueError(
-                    'The wildcard string "*" may only be passed to allow_origins as a '
-                    'string literal, not inside an iterable.'
-                )
+            if allow_origins == '*':
+                self.allow_origins_wildcard = True
+            else:
+                if isinstance(allow_origins, str):
+                    allow_origins = [allow_origins]
+                allow_origins = frozenset(allow_origins)
+                if '*' in allow_origins:
+                    raise ValueError(
+                        'The wildcard string "*" may only be passed to allow_origins as a '
+                        'string literal, not inside an iterable.'
+                    )
+            self.allow_origins_filter = self.create_origin_filter(allow_origins)
 
         if expose_headers is not None and not isinstance(expose_headers, str):
             expose_headers = ', '.join(expose_headers)
         self.expose_headers = expose_headers
 
-        if allow_credentials is None:
-            allow_credentials = frozenset()
-        elif allow_credentials != '*':
-            if isinstance(allow_credentials, str):
-                allow_credentials = [allow_credentials]
-            allow_credentials = frozenset(allow_credentials)
-            if '*' in allow_credentials:
-                raise ValueError(
-                    'The wildcard string "*" may only be passed to allow_credentials '
-                    'as a string literal, not inside an iterable.'
-                )
-        self.allow_credentials = allow_credentials
+        if isinstance(allow_credentials, Callable):
+            self.allow_credentials_filter = allow_credentials
+        else:
+            if allow_credentials is None:
+                allow_credentials = frozenset()
+            elif allow_credentials != '*':
+                if isinstance(allow_credentials, str):
+                    allow_credentials = [allow_credentials]
+                allow_credentials = frozenset(allow_credentials)
+                if '*' in allow_credentials:
+                    raise ValueError(
+                        'The wildcard string "*" may only be passed to allow_credentials '
+                        'as a string literal, not inside an iterable.'
+                    )
+            self.allow_credentials_filter = self.create_origin_filter(allow_credentials)
+
+    def create_origin_filter(self, allow_list):
```

Comment:
Should this be part of the public interface? It seems like it should be private (prefixed w/ an undescore), at least in the first release until we discover otherwise.